### ISSUE TITLE
`compute_rank_mahalanobis` ignores index argument

### ISSUE THREAD

**Issue by benthestatistician** [markmfredrickson/optmatch#128]
...in the event that there's an exact matching constraint in force. The function presently computes and returns a distance for each treatment/control combination, even when provided a list of specific treatment-control matching possibilities for which a distances is needed.   

to do's:

- [ ] If possible, fix it in compute_rank_mahananobis
- [x] If not possible, workaround on R side 
- [x] Add a test case to test.rank.mahal.R

**Comment by benthestatistician** [comment]
In the example in #122 , we have 

    > length(within@rows)
    [1] 4889857

whereas `compute_rank_mahalanobis` has give a result of length 

   > length(dists)
   [1] 22887076

matching the product of the overall number of rows and columns. 

    > within@dimension
   [1]  2252 10163

**Comment by benthestatistician** [comment]
@jwbowers since you've caught the rank_mahalanobis bug, would you be willing to help out w/ writing of test cases?  

I figured I'd start by writing out some tests where you do a rank_mahalanobis calculation by pre-computing the ranks in R, then applying ordinary Mahalanobis. That led me to try the following, which does not work, but which is more likely to represent a misunderstanding on my part of what rank_mahalanobis is trying to do than a problem with the code.  I.e., maybe you can help me improve it w/o much effort.  

```{r}
test_that("compute_rank.mahal results match ordinary Mahalanobis's", {
   nr <- 10L
   z <- integer(nr)
    z[sample(1:nr, nr / 2L)] <- 1L

   X <- as.matrix(1L:nr)
   df <- data.frame(z = z, X)
   expect_equivalent(match_on(z~., data=df, method="rank_mahalanobis"),
                     match_on(z~., data=df, method="mahalanobis")
                     )
}
)
```

(Once I get something like this going, should be easy to write test cases for the combo of exact matching and rank Mahalanobis.)

**Comment by benthestatistician** [comment]
In tests/testthat/test.rank.mahal.R, `compute_rank_mahalanobis` is tested against the rank Mahalanobis function that Rosenbaum has posted to his site. I confirmed that their results match up in this example also, even as they differ from those of `compute_mahalanobis`. 

The likely cause of the  difference is underlying differences in how the covariance is being computed.  optmatch has always calculated the covariances needed for Mahalanobis distance computations by pooling covariances in treatment and control groups.  In defining rank-based Mahalanobis, Rosenbaum uses an unpooled covariance.  Anyway there's no indication of anything amiss, other than the problem with `compute_rank_mahalanobis` noted at the top of this ticket. We'll need another strategy for test cases.

**Comment by kkbrum** [comment]
@benthestatistician FYI, this "ignoring" the "within" argument during the actual computation process is causing issues for the new version of student matching for ADSY (since the data is really too large when "within" is ignored). It won't help me (I'll work around it), but I would advocate for actually fixing this issue. At the very least, a note in the "match_on" documentation page about the futility of "within" for "method = rank_mahalanobis" would be nice.

**Comment by benthestatistician** [comment]
Thanks for taking the trouble to track this down and post a comment on it, @kkbrum.  I think I resorted to a shortcut because I wasn't confident I'd be able to update @nullsatz 's code in src/r_smahal.* and src/smahal.*.   

Now that we've got Josh back on the team, though: Josh, could you look into enabling the index argument of the rank mahanalobis computation, so that when a BISM is requested it only has to compute and store the within-block distances?  My workaround was to use the existing rank mahalanobis routines to compute a distance for each treatment control pair.  In sufficiently large matching problems this is no good.

**Comment by nullsatz** [comment]
I was working on making the smallest change possible.

Simply replacing the C code version,
```sqrt(r_smahal(NULL, data, z))```,
with the `Rcpp` version that does support an index argument,
```compute_mahalanobis(index, apply(data, 2, rank), z)```,
and I don't know why.

```
Failure (test.rank.mahal.R:74:5): compute_rank.mahal returns results similar to the Rosenbaum code
as.matrix(match_on(z ~ X, method = "rank")) not equivalent to `reference_rankmahal`.
25/25 mismatches (average diff: 0.658)
[1] 4.45 - 3.49 == 0.962
[2] 4.24 - 3.45 == 0.790
[3] 5.57 - 3.54 == 2.034
[4] 3.64 - 3.07 == 0.568
[5] 4.38 - 3.19 == 1.191
[6] 4.34 - 3.78 == 0.561
[7] 4.28 - 3.89 == 0.388
[8] 5.42 - 3.93 == 1.481
[9] 3.85 - 3.68 == 0.178
...
```

**Comment by kkbrum** [comment]
The rank-based mahalanobis distance also rescales the ranked variables before applying the mahalanobis distance calculation.

**Comment by nullsatz** [comment]
> The rank-based mahalanobis distance also rescales the ranked variables before applying the mahalanobis distance calculation.

So I went back and looked at my C source code and did
```
scale(apply(data, 2, rank, ties.method="average"),
          center = TRUE, scale = FALSE)
```
but that didn't seem to change anything.

I guess I don't really know what 'rescales the ranked variables' means.

**Comment by benthestatistician** [comment]
1. I may be missing something, but if rescaling the rank-transformed variables means something like scaling them to lie in [0,1], or centering before scaling by a reciprocal SD, then I wouldn't expect that to meaningfully change the calculation. 
2. In [my comment above dated July, 8 2017](https://github.com/markmfredrickson/optmatch/issues/128#issuecomment-313880210) I chalk up differences between my expectations for a "rank-based Mahalanobis distance" vs distances from Paul Rosenbaum's implementation to mine calling for the use of a pooled covariance vs Paul's using unpooled. This would also explain why Josh B's "smallest possible change to Mahalanobis" proposal gives a result different than what our tests expect: the tests are calibrate to agree with Paul's version. 
3. It's good to be able to offer a fast and convenient rank-based Mahalanobis calculation that agrees with what Paul R offers, but it would also be nice to have a fast and performant rank-based Mahalanobis that uses pooled rather than unpooled covariance -- i.e., Josh's "smallest possible change to Mahanalobis" version. Perhaps via "compute_ranked_mahalanobis_pooled()"
4. A first unit test for such a thing might be the "compute_rank.mahal results match ordinary Mahalanobis's" test code given in [my comment above dated July 6, 2017](https://github.com/markmfredrickson/optmatch/issues/128#issuecomment-313494929).  
5. Given its potential utility for @kkbrum's current work, I'd put `compute_rank_mahalanobis_pooled()` ahead of the project of improving the use of indices with the current `compute_rank_mahalanobis()` in the priority queue.

**Comment by nullsatz** [comment]
Thanks @kkbrum and @benthestatistician for the info. I've cobbled together a minimal feature. Here it is in action.

```
test_that(
    "compute_rank_mahalanobis_pooled results match ordinary Mahalanobis's",
    {
    nr <- 10L
    z <- integer(nr)
    z[sample(1:nr, nr / 2L)] <- 1L

    X <- as.matrix(1L:nr)
    df <- data.frame(z = z, X)
    expect_equivalent(match_on(z~., data=df, method="pooled_rank"),
                      match_on(z~., data=df, method="mahalanobis"))
    })
```

Not sure how to say this in English but `pmatch` wants the front end to be distinct rather than the rear end. So I put pooled in front. Sorry, Ben.

Here is the very humble new function.

```
compute_rank_mahalanobis_pooled <- function(index, data, z) {
    if (!all(is.finite(data))) {
        stop("Infinite or NA values detected in data for Mahalanobis computations.")
    }

    if (is.null(index)) return(sqrt(r_smahal(NULL, data, z)))

    if (is.null(rownames(data)) | !all(index %in% rownames(data)))
        stop("data must have row names matching index")

    return(compute_mahalanobis(index, apply(data, 2, rank), z))
}
```

This all passes for me except for an unrelated error involving `survival::strata`.

I'm working in the branch [issue-128-rank-mahal-index](https://github.com/markmfredrickson/optmatch/tree/issue-128-rank-mahal-index)

**Comment by benthestatistician** [comment]
Thanks for the quick work!  @kkbrum can now pull from the  [issue-128-rank-mahal-index](https://github.com/markmfredrickson/optmatch/tree/issue-128-rank-mahal-index) branch for pooled covariance rank Mahalanobis distances on a designated subset of the treatment-control pairs.   That could happen in parallel with a few further adjustments:
1. It seems that `r_smahal()` will perform its variance calculation without pooling of covariances, rather than with.  Could the `is.null(index)` arm be adjusted to also pool its covariance of the rank-transformed covariates, as `compute_mahanalobis()` does? 
2. I'm OK with "pooled" coming first in "`pooled_rank`" , but I think I'd prefer "`pooled_cov_rank_mahalanobis`".  Assuming unambiguous abbreviations are OK, eg. "`pooled_cov_rank_mahal`" or even "`pooled`"  -- your mention of `pmatch()` makes me suspect they are. 
3. The "`compute_`" function can be renamed to "`compute_pooled_cov_rank_mahalanobis()`", or whatever.

**Comment by nullsatz** [comment]
1. I removed the `if(is.null(index))` and associated block. I don't think it's necessary.

(2) and (3) are done now.

**Comment by benthestatistician** [comment]
Could you add tests excercising it with and without a within= argument to `match_on()`, @nullsatz, to verify that's being properly handled? And any other unit tests that seem to make sense.   I added a NEWS entry and some docs updates (c851969), so we'll be about ready to merge after the tests are filled out.

**Comment by nullsatz** [comment]
I added a [line](https://github.com/markmfredrickson/optmatch/blob/841b5b0854b8a5216fc05969fb9befaff7169201/tests/testthat/test.match_on.R#L177) in one of the `match_on` tests.

I also added a quick `within` argument (of a style I saw on lots of other tests) in the `rank.mahal` tests [here](https://github.com/markmfredrickson/optmatch/blob/841b5b0854b8a5216fc05969fb9befaff7169201/tests/testthat/test.rank.mahal.R#L139).

**Comment by kkbrum** [comment]
After some more exploration, I believe the mismatches that Ben and Josh have both noticed compared to Rosenbaum's robust mahalanobis reference are not just due to pooled vs unpooled covariances (sure, that is part of it, but not the whole story). 

I bring this up because switching from the version of robust mahalanobis distances I was temporarily using (based on MatchIt's transform_covariates function) to the new compute_rank_mahalanobis_pooled caused worse performance in terms of match balance. I modified my original version to use pooled covariance to see if that was the issue, but that did not bring the distances into alignment with compute_rank_mahalanobis_pooled. 

I did some more digging and pulled up Rosenbaum's code for DOS2::smahal. I believe the piece missing from compute_rank_mahalanobis_pooled is the part where (quote taken from the smahal documentation): "the variances of the ranked columns are rescaled so they are all the same, all equal to the variance of untied ranks": 

`cv <- stats::cov(X)`
 `vuntied <- stats::var(1:n)`
 `rat <- sqrt(vuntied/diag(cv))`
`cv <- diag(rat) %*% cv %*% diag(rat)`

Given that I have a whole bunch of binary covariates in my distance, differing in their levels of rarity and thus variance, I believe this would explain the worse performance.

I didn’t have time to look carefully as to whether this is being done somewhere in
compute_rank_mahalanobis_pooled, but from what I’ve seen, I don’t believe so.

**Comment by benthestatistician** [comment]
Thanks much for this, Katherine.  Now I see what you had in mind earlier about the rescaling. 

One think I can say is that I recall seeing reference to this in a comment in @nullsatz's C code from 10 years ago.  I think you're right that it's not in `compute_rank_mahalanobis_pooled()` now, but now there are a couple ways to put it there, your/Paul's R implementation and Josh's earlier C implementation.

**Comment by nullsatz** [comment]
I implemented the requested changes. One problem I ran into is that `rat` must be wrapped in `as.matrix` for edge cases.

**Comment by benthestatistician** [comment]
Thanks!  I like the well-named additions `scale_addressing_ties()` and `save_invert()`,  which nicely improve readability of the code.  How about adding some quick unit tests verifying in particular cases that:

- [x]  given a pair of columns of integer ranks of some data including a few ties,  `scale_addressing_ties()` indeed returns a matrix with diagonal entries all equal to `var(1:n)`
- [x] given a full rank symmetric square matrix, `safe_invert()` returns the same thing as `solve()`
- [x] given a rank-deficient symmetric square matrix, `safe_invert()` returns the same thing as `MASS::ginv()`? 

~~Then I'd say this is ready for a linked PR.~~ Also,

- [x] write an `index` arg respecting `compute_rank_mahananobis()` (!).  Like `compute_pooled_cov_rank_mahalanobis()`, but with a cov of all the observations. 

(Then we'll be ready for the PR.)

**Comment by nullsatz** [comment]
I got back to my office and I'm still running into lots of `non-conformable arguments` errors. This is the current code.
```
scale_addressing_ties <- function(n, cv) {
    cv <- as.matrix(cv)
    vuntied <- var(1:n)
    rat <- as.matrix(sqrt(vuntied/diag(cv)))
    print('')
    print('cv')
    print(cv)
    print('rat')
    print(rat)
    print('diag(rat)')
    print(diag(rat))
    return(diag(rat) %*% cv %*% diag(rat))
}
```

Here is one bad thing that happens.
```
[1] ""
[1] "cv"
           [,1]       [,2]
[1,] 1.73333333 0.08888889
[2,] 0.08888889 1.37777778
[1] "rat"
         [,1]
[1,] 2.299666
[2,] 2.579385
[1] "diag(rat)"
[1] 2.299666
```
So I'm not going to be able to do `diag(rat) %*% cv %*% diag(rat)`

In this case, should the return value be `diag(rat) * cv * diag(rat)` ?

Wait! Is rat supposed to be a diagonal matrix with the same dim as cv? I will see if I can get that to work.

**Comment by benthestatistician** [comment]
As defined in the original code we're working from, `rat` is supposed to be a vector (w/ no `dim` attribute), length equal to the row or column extent of `cv`.  Then `diag(rat)` is a diagonal matrix with the same dim as `cv`.  

I think you'd get this in the code above by switching out
```r
rat <- as.matrix(sqrt(vuntied/diag(cv)))
```
in favor of 
```r
rat <- sqrt(vuntied/diag(cv))
```

**Comment by nullsatz** [comment]
Ok, thanks for the info. That helped clear up a lot. After that, I just had to handle the case where rat has length 1. The function `scale_addressing_ties` and its unit test is working now.

**Comment by nullsatz** [comment]
I cleaned up the code somewhat, and I'm seeing a failing unit test that I wrote.

```
Failure (test.rank.mahal.R:136:5): compute_pooled_cov_rank_mahalanobis results match ordinary Mahalanobis's
match_on(z ~ ., data = df, method = "pooled_cov") not equivalent to match_on(z ~ ., data = df, method = "mahalanobis").
Mean relative difference: 0.29986
```

This was being hidden because at the end of the `compute_pooled_cov...`  function I had a call to `compute_mahalanobis` instead of `mahalanobis_helper`.

**Comment by nullsatz** [comment]
unit tests for the new utility functions are in place and passing

**Comment by benthestatistician** [comment]
> This was being hidden because at the end of the `compute_pooled_cov...` function I had a call to `compute_mahalanobis` instead of `mahalanobis_helper`.

So the wrong computation gave the expected answer but the fixed calculation doesn't?  One has to suspect the test is asserting a mistaken expectation; at a glance I think that's possible.

**Comment by nullsatz** [comment]
I wonder how we can test `compute_pooled_cov_rank_mahalanobis` then.

**Comment by benthestatistician** [comment]
We ought to be able to modify that test to get it to work. I think that
should happen if on their own (prior to rank transformation) the Xes' two
with-group variances are the same.  It may also be necessary that the two
sets of within-group Xes each avoid ties and occupy an evenly spaced grid,
e.g. (2,4,..10) or (1, 3, ..., 9).  Perhaps there are additional
requirements that aren't occurring to me.

**Comment by kkbrum** [comment]
No idea if this is related to the test you're discussing or not, but backing up a bit, @nullsatz currently you are invoking "scale_addressing_ties" within the cov calculation for treated & control separately before pooling. If I'm not mistaken, I believe that the scaling should be addressed just once after the pooling has already occurred.

**Comment by benthestatistician** [comment]
Katherine, your choice of words suggests you have a refererence in mind for the "pooled covariance rank Mahalanobis" concept -- do you? I figured this was something we got to make up as we go along, Rosenbaum only ever discussing the version using a full-sample covariance. 

If there is a published reference then I imagine we'd want to follow it.  If no, I can see it either way, but the way we've got it now makes a little more sense to me.  (I recognize that the point of your comment may be that the other way makes more sense to you -- happy to compare notes further or offline, depending on what you recommend.)

**Comment by kkbrum** [comment]
No reference. I was mostly basing this off the observation that the current code yields some very extreme distances that I found bizarre (1.5% of distances for me are in the 3000's), vs if you do the scaling after pooling, as I had imagined/tried, this didn't happen. That being said, the match quality comes out absolutely terrible (worse than before matching) if you scale after pooling, vs it comes out excellent as written, with scaling before pooling. So I rescind my comment; apologies!

**Comment by benthestatistician** [comment]
I think this is ready for for linked PR

**Comment by benthestatistician** [comment]
Re the currently failing test of `compute_pooled...()`, I'd next try modifying the following lines of the `compute_pooled...()` def, as well as the similar lines just below them modifying `mc`:
```r
      mt <- cov(treated) * (sum(z) - 1) / (length(z) - 2)
      mt <- scale_addressing_ties(nt, mt)
```
The mod I suggest is to apply `scale_addressing_ties()` before bringing in the `(sum(z) - 1) / (length(z) - 2)` multiplier, not after.

**Comment by nullsatz** [comment]
Here's what the [changed code](https://github.com/markmfredrickson/optmatch/blob/68afe5eaabc32220a98aaaed21469c55719a5ca6/R/match_on.R#L573) looks like
```
      nt <- nrow(treated)
      mt <- scale_addressing_ties(nt, cov(treated))
      mt <- mt * (sum(z) - 1) / (length(z) - 2)
```
`mc` is computed similarly.


Interestingly, now the mean relative distance is 0.5
```
Failure (test.rank.mahal.R:141:5): compute_pooled_cov_rank_mahalanobis results match ordinary Mahalanobis's
match_on(z ~ ., data = df, method = "pooled_cov") not equivalent to match_on(z ~ ., data = df, method = "mahalanobis").
Mean relative difference: 0.5
```

**Comment by nullsatz** [comment]
I loosened the criteria for passing. Now results from match_on using compute_pooled_rank_mahalanobis need only be a scalar multiple of results from match_on using just mahalanobis.

I had to add MASS to `Suggests` in the `DESCRIPTION` file because one of my tests uses `ginv` from `MASS`

I also had to add an `#' @importFrom stats var` for one of the new functions in `utilities.R`.

The above two changes were to get `R CMD check` to pass.

The tests now pass.
